retrying otp does not call onboarding

srieteja opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or


Please provide the following information:
-->
closes https://github.com/atsign-foundation/at_java/issues/14
**- What I did**

- Solved a bug where re-entering otp does not call onboarding
- Handled a case when user exhausts their quota of free atsigns by displaying appropriate message
- Also restructure config yaml and changed the config variables to camelCase

**- How I did it**

- reposition if-else blocks so that the block calling onboarding class is executed every time
- calling API to check if they have exhausted free atsign quota before trying to register it to a users email

**- How to verify it**

- use register CLI and enter incorrect otp and then enter correct otp. Expected behaviour is that onboarding is successfully completed
- try registering a new atsign with an email that has exhausted free atsign limit, the CLI should throw an exception "free atsign limit exhausted"

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

srieteja committed:
reposition if-else blocks

gkc commented on at_client/src/main/java/org/atsign/client/cli/Register.java:
```diff
@@ -24,16 +24,16 @@ public static void main(String[] args) throws Exception {
+        String rootDomain = configReader.getProperty("rootDomain");
+        String rootPort = configReader.getProperty("rootPort");
+        String registrarUrl = configReader.getProperty("registrarUrl");
+        String apiKey = configReader.getProperty("apiKey");
```
This is a breaking change. If the property is null when retrieving with the new name, then try to retrieve using the old name

Consider also adding some structure to the yaml; now's the best time to do this before it gets fossilized by usage

e.g.

root:
  domain: "xxxx"
  port: nnn
registrar:
  url: "https://..."
  apiKey: "xyz456abc123..."

gkc commented on at_client/src/main/resources/config.yaml:
```diff
@@ -1,4 +1,4 @@
+apiKey: "477b-876u-bcez-c42z-6a3d" ##staging api key. please obtain a key to the production registrar api and replace it
+rootDomain: "root.atsign.wtf" ##staging root server. use "root.atsign.org" to use the production root domain
+rootPort: "64" 
+registrarUrl: "https://my.atsign.wtf/api/app/v2" ##staging registrar url. please use "https://my.atsign.com/api/app/v2" to use the production registrar api
```
See earlier comment about backwards compatibility / breaking change.

srieteja replied:
Restructuring the yaml required me to modify the mechanism/method that reads the yaml file. ConfigReader.getProperty() now accepts different parameters than it did previously. This might not allow retrieving a configuration using the old name.

srieteja committed:
structuring config yaml

srieteja committed:
handling case where user exhausts atsign limit

srieteja commented:
@gkc the latest commit handles the case when a user exhausts their free atsign limit. The current behaviour is that the CLI prints a log saying "max free atsigns reached". In this case do we want to add functionality where the user could choose from their existing atsigns and the Register CLI would onboard it for them ?

gkc commented on at_client/src/main/java/org/atsign/client/cli/Register.java:
```diff
@@ -24,16 +24,16 @@ public static void main(String[] args) throws Exception {
-        String apiKey = configReader.getProperty("API_KEY");
-        
-        if (rootDomain == null || rootPort == null || registrarUrl == null || apiKey == null){
+        String rootDomain = configReader.getProperty("rootServer", "domain");
```
Add here, for backwards compatibility 
```
if (rootDomain == null) {
  rootDomain = configReader.getProperty("ROOT_DOMAIN");
}
```
and similar for each of the next three config properties

gkc commented on at_client/src/main/java/org/atsign/client/util/RegisterUtil.java:
```diff
@@ -108,34 +109,53 @@ public Boolean registerAtsign(String email, AtSign atsign, String registrarUrl,
      */
-    public String validateOtp(String email, AtSign atsign, String otp, String registrarUrl, String apiKey)
+    public String validateOtp(String email, AtSign atsign, String otp, String registrarUrl, String apiKey,
+            boolean confirmation)
```
This is a breaking change. Please make `confirmation` an optional parameter. Customer is directly using this method.

gkc commented on at_client/src/main/java/org/atsign/client/util/RegisterUtil.java:
```diff
@@ -108,34 +109,53 @@ public Boolean registerAtsign(String email, AtSign atsign, String registrarUrl,
      * @throws IOException
```
Please add documentation explaining use of the new 'confirmation' parameter (which should be optional as mentioned in next comment)

gkc commented on at_client/src/main/java/org/atsign/config/ConfigReader.java:
```diff
@@ -13,13 +13,15 @@
+    private Map<String,String>propertyMap;
 
-    public String getProperty(String property) throws StreamReadException, DatabindException, FileNotFoundException{
+    public String getProperty(String property, String subProperty) throws StreamReadException, DatabindException, FileNotFoundException{
```
Please preserve the getProperty(String property) for backwards compatibility and either add a new method, getNestedProperty(String property, String subProperty) or getNestedProperty(List<String> propertyNames) or similar

gkc commented on at_client/src/main/resources/config.yaml:
```diff
@@ -1,4 +1,7 @@
+
+registrar:
+  url: "https://my.atsign.wtf/api/app/v2" ##staging registrar url. please use "https://my.atsign.com/api/app/v2" to use the production registrar api
+  apiKey: "477b-876u-bcez-c42z-6a3d" ##staging api key. please obtain a key to the production registrar api and replace it
```
Please re-add the original config names, specifying in comments that they are there for backwards compatibility and that users should use, for example, rootServer:domain rather than ROOT_DOMAIN as ROOT_DOMAIN will be deprecated in future

srieteja committed:
addressing review comments

srieteja commented:
latest commit addresses all review comments. Although I'm stuck with an issue where API sends an ArrayList of atsigns and I cannot cast into string or any other format.

gkc commented on at_client/src/main/java/org/atsign/client/util/RegisterUtil.java:
```diff
@@ -100,42 +109,86 @@ public Boolean registerAtsign(String email, AtSign atsign, String registrarUrl,
+            boolean ... confirmation)
             throws IOException, AtException {
+        //setting default confirmation to true in case it's not provided
+        Boolean defaultConfirmation = (confirmation.length == 0) ? true : confirmation[0];
```
This is not what I was looking for. I asked for the confirmation parameter to be made optional, I was thinking as a named optional parameter e.g. 
`public String validateOtp(String email, AtSign atsign, String otp, String registrarUrl, String apiKey, {boolean confirmation = true})`

gkc replied:
OMG I'm thinking of Dart

srieteja replied:
yes, as far as StackOverflow goes this is one of the limited ways we can do optional parameters

gkc replied:
Ignore my previous comments

To preserve backwards compatibility
1. Add new method `public String validateOtp(String email, AtSign atsign, String otp, String registrarUrl, String apiKey, boolean confirmation)`
2. Keep existing method `public String validateOtp(String email, AtSign atsign, String otp, String registrarUrl, String apiKey)` with implementation `{validateOtp(email, atsign, otp, registrarUrl, apiKey, true); }` assuming that the default value for 'confirmation' is true

gkc commented:
@srieteja 
> Although I'm stuck with an issue where API sends an ArrayList of atsigns and I cannot cast into string or any other format.

Can you explain more how you're stuck?

srieteja commented:
The API responds with different formats for different requests. I have been able to handle using two collection types Map<String, String> and Map<String, Map<String, String>>. But for the added case now the API responds with an ArrayLIst as one of the parameters instead of String. If I try to print that ArrayList through sysout I'm getting a run-time error that says ArrayList cannot be casted to String and if I try to extract that into a variable of type ArrayList so that I can handle I'm getting a compile-time error that won't allow me to assign the list from API response to the variable as I have specified the type as Map<String, String>. I have tried to use Dynamic instead of String that has been giving me an error too for some apparent reason.

gkc commented:
> The API responds with different formats for different requests. I have been able to handle using two collection types Map<String, String> and Map<String, Map<String, String>>. But for the added case now the API responds with an ArrayLIst as one of the parameters instead of String. If I try to print that ArrayList through sysout I'm getting a run-time error that says ArrayList cannot be casted to String and if I try to extract that into a variable of type ArrayList so that I can handle I'm getting a compile-time error that won't allow me to assign the list from API response to the variable as I have specified the type as Map<String, String>. I have tried to use Dynamic instead of String that has been giving me an error too for some apparent reason.

Please explain in the code, with commented code if necessary, so I can checkout the branch, understand better the problem, and propose a solution.

srieteja committed:
throws castException

srieteja commented:
[response.zip](https://github.com/atsign-foundation/at_java/files/8780395/response.zip)
In the above zip is the response json for the case I'm facing the error. In RegisterUtil.java line 175 I'm extracting the json value of field "data" to a Map<String, String>. Please try running the latest commit which most likely will throw an error due to newly added line 184(RegisterUtil.java)

srieteja committed:
final commit

srieteja committed:
method overloading validateOtp

srieteja commented:
@gkc I was able to solve the thing I was stuck with. If the PR looks good to you we can go ahead and merge it.

srieteja committed:
readme updated

srieteja committed:
udpated java docs

gkc reviewed (approved):
LGTM
